File: src/34. Find First and Last Position of Element in Sorted Array/main.ts
// Solved by @kitanoyoru
// https://leetcode.com/problems/find-first-and-last-position-of-element-in-sorted-array/

const searchRange = (nums: number[], target: number): number[] => {
  let start = 0,
    end = nums.length - 1,
    mid = 0,
    ans= new Array(2)


  while (start <= end) {
    mid = start + Math.floor((end - start) / 2)
    if (nums[mid] == target) {
      ans[0] = mid
      end = mid - 1
    } else if (nums[mid] < target) {
      start = mid + 1
    } else {
      end = mid - 1
    }
  }
  
  start = 0;
  end = nums.length - 1;

  while (start <= end) {
    mid = start + Math.floor((end - start) / 2)
    if (nums[mid] == target) {
      ans[1] = mid
      start = mid + 1
    } else if (nums[mid] < target) {
      start = mid + 1
    } else {
      end = mid - 1
    }
  }

  return ans.includes(undefined) ? [-1, -1] : ans;
}

console.log(searchRange([5, 7, 7, 8, 8, 10], 6))

File: src/34. Find First and Last Position of Element in Sorted Array/main.go
func searchRange(nums []int, target int) []int {
  start := binarySearch(nums, target, true)
  
  if start == -1 {
    return []int {-1, -1}
  }

  end := binarySearch(nums, target, false)

  return []int {start, end}
}

func binarySearch(nums []int, target int, isFindingStart bool) int {
  left, right, ans := 0, len(nums) - 1, -1

  for left <= right {
    mid := left + (right - left) / 2

    if nums[mid] == target {
      if ans == -1 {
        ans = mid
      } else {
        if isFindingStart {
          if mid < ans {
            ans = mid
          }
        } else {
          if mid > ans {
            ans = mid
          }
        }
      }

      if isFindingStart {
        right = mid - 1
      } else {
        left = mid + 1
      }
    } else if nums[mid] > target {
      right = mid - 1
    } else {
      left = mid + 1
    }
  }

  return ans
}
